checker/env.go

```diff
@@ -57,20 +58,49 @@ func (e *Env) Add(decls ...*checked.Decl) {
 	}
 }
 
-func (e *Env) addFunction(decl *checked.Decl) {
-	current := e.declarations.FindFunction(decl.Name)
-	if current != nil {
-		if current.Name != decl.Name {
+func (e *Env) addOverload(f *checked.Decl, overload *checked.Decl_FunctionDecl_Overload) {
+	function := f.GetFunction()
+	emptyMappings := types.NewMapping()
+	overloadFunction := types.NewFunction(overload.GetResultType(),
+		overload.GetParams()...)
+	overloadErased := types.Substitute(emptyMappings, overloadFunction, true)
+	for _, existing := range function.GetOverloads() {
+		existingFunction := types.NewFunction(existing.GetResultType(),
+			existing.GetParams()...)
+		existingErased := types.Substitute(emptyMappings, existingFunction, true)
+		overlap := (types.IsAssignable(emptyMappings, overloadErased, existingErased) != nil ||
+			types.IsAssignable(emptyMappings, existingErased, overloadErased) != nil)
+		if overlap &&
+			overload.GetIsInstanceFunction() == existing.GetIsInstanceFunction() {
+			e.errors.overlappingOverload(common.NoLocation, f.Name, overloadFunction,
```

Comment:
Consider introducing decls.FormatDecl() to go along with types.FormatType(). Formatting a type results in a human-friendly representation of a proto-based type, so we should have equivalently human-friendly representations of declarations.

Reply:
maybe decls.FormatDecl() should be done in separate change because we are not using this in this change ? We are passing function types to the error which use types.FormatType to print the function type.

Reply:
Printing a function from FormatType appears to be missing from types.go. Please add it and ignore my earlier suggestion of decls.FormatDecl().

I would expect the function type listed in the error message to be formatted as follows:
```
<input>:-1:0: overlapping overload for name  'myfunc' (function(dyn) : dyn) ' 
cannot be distinguished from '(function(dyn) : dyn)'
```

Also, this error message should include the overload id of the candidate and the id of the one it collides with.

Reply:
I added formatFunction. Let me know if the formatting of the error message is OK now.